### PR TITLE
EN-11730: use a context + predefined channel for get all leaves operations

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
Refactor GetAllLeaves / GetAllLeavesOnChannel functions so they receive the channel instead of creating it + they receive a context as well that can trigger an early exit on the processing.

Implemented a timeout mechanism for node's API endpoints that called that function

### CODE REVIEW

File: process/peer/process.go [hunk 1]
```diff
@@ -572,7 +573,8 @@ func (vs *validatorStatistics) GetValidatorInfoForRootHash(rootHash []byte) (map
 		log.Debug("GetValidatorInfoForRootHash", sw.GetMeasurements()...)
 	}()
 
-	leavesChannel, err := vs.peerAdapter.GetAllLeaves(rootHash)
+	leavesChannel := make(chan core.KeyValueHolder, common.TrieLeavesChannelDefaultCapacity)
+	err := vs.peerAdapter.GetAllLeaves(leavesChannel, context.Background(), rootHash)
```
Comment (iulianpascalau) [inline process/peer/process.go:577]: :+1: 
`context.Background()` is correct in this situation

File: state/syncer/userAccountsSyncer.go [hunk 2]
```diff
@@ -211,7 +211,8 @@ func (u *userAccountsSyncer) syncAccountDataTries(
 		return err
 	}
 
-	leavesChannel, err := mainTrie.GetAllLeavesOnChannel(mainRootHash)
+	leavesChannel := make(chan core.KeyValueHolder, common.TrieLeavesChannelDefaultCapacity)
+	err = mainTrie.GetAllLeavesOnChannel(leavesChannel, context.Background(), mainRootHash)
```
Comment (iulianpascalau) [inline state/syncer/userAccountsSyncer.go:215]: :+1: 
`context.Background()` is correct in this situation

File: update/genesis/export.go [hunk 3]
```diff
@@ -269,7 +270,8 @@ func (se *stateExport) exportTrie(key string, trie common.Trie) error {
 		return err
 	}
 
-	leavesChannel, err := trie.GetAllLeavesOnChannel(rootHash)
+	leavesChannel := make(chan core.KeyValueHolder, common.TrieLeavesChannelDefaultCapacity)
+	err = trie.GetAllLeavesOnChannel(leavesChannel, context.Background(), rootHash)
```
Comment (iulianpascalau) [inline update/genesis/export.go:274]: :+1: 
`context.Background()` is correct in this situation

File: node/trieIterators/delegatedListProcessor_test.go [hunk 4]
```diff
@@ -174,19 +175,17 @@ func createDelegationScAccount(address []byte, leaves [][]byte, rootHash []byte)
...
+				go func() {
+					for _, leafBuff := range leaves {
+						leaf := keyValStorage.NewKeyValStorage(leafBuff, nil)
+						ch <- leaf
+					}
 
-				close(ch)
-			}()
+					close(ch)
+				}()
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply node/trieIterators/delegatedListProcessor_test.go:186]: done

File: state/accountsDB_test.go [hunk 5]
```diff
@@ -1304,21 +1305,21 @@ func TestAccountsDB_GetAllLeaves(t *testing.T) {
 
 	getAllLeavesCalled := false
 	trieStub := &trieMock.TrieStub{
-		GetAllLeavesOnChannelCalled: func(rootHash []byte) (chan core.KeyValueHolder, error) {
-			getAllLeavesCalled = true
+		GetAllLeavesOnChannelCalled: func(ch chan core.KeyValueHolder, ctx context.Context, rootHash []byte) error {
+				getAllLeavesCalled = true
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply state/accountsDB_test.go:1309]: fixed

File: trie/errors.go [hunk 6]
```diff
@@ -105,3 +105,6 @@ var ErrKeyNotFound = errors.New("key not found")
 
 // ErrNilEpochNotifier signals that the provided EpochNotifier is nil
 var ErrNilEpochNotifier = errors.New("nil EpochNotifier")
+
+// ErrContextDeadlineExceeded signals that a context has exceeded it's deadline
+var ErrContextDeadlineExceeded = errors.New("context deadline exceeded")
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply trie/errors.go:110]: removed

File: node/trieIterators/delegatedListProcessor.go [hunk 7]
```diff
@@ -124,7 +126,8 @@ func (dlp *delegatedListProcessor) getDelegatorsList(delegationSC []byte) ([][]b
 		return nil, fmt.Errorf("%w for delegationSC %s", err, hex.EncodeToString(delegationSC))
 	}
 
-	chLeaves, err := delegatorAccount.DataTrie().GetAllLeavesOnChannel(rootHash)
+	chLeaves := make(chan core.KeyValueHolder, common.TrieLeavesChannelDefaultCapacity)
+	err = delegatorAccount.DataTrie().GetAllLeavesOnChannel(chLeaves, context.Background(), rootHash)
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply node/trieIterators/delegatedListProcessor.go:130]: done

File: node/trieIterators/directStakedListProcessor.go [hunk 8]
```diff
@@ -51,7 +53,8 @@ func (dslp *directStakedListProcessor) getAllStakedAccounts(validatorAccount sta
 		return nil, err
 	}
 
-	chLeaves, err := validatorAccount.DataTrie().GetAllLeavesOnChannel(rootHash)
+	chLeaves := make(chan core.KeyValueHolder, common.TrieLeavesChannelDefaultCapacity)
+	err = validatorAccount.DataTrie().GetAllLeavesOnChannel(chLeaves, context.Background(), rootHash)
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply node/trieIterators/directStakedListProcessor.go:57]: done

File: node/trieIterators/stakeValuesProcessor.go [hunk 9]
```diff
@@ -93,7 +95,8 @@ func (svp *stakedValuesProcessor) computeBaseStakedAndTopUp() (*big.Int, *big.In
 	}
 
 	// TODO investigate if a call to GetAllLeavesKeysOnChannel (without values) might increase performance
-	chLeaves, err := validatorAccount.DataTrie().GetAllLeavesOnChannel(rootHash)
+	chLeaves := make(chan core.KeyValueHolder, common.TrieLeavesChannelDefaultCapacity)
+	err = validatorAccount.DataTrie().GetAllLeavesOnChannel(chLeaves, context.Background(), rootHash)
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply node/trieIterators/stakeValuesProcessor.go:99]: done

**Review by gabi-vuls — APPROVED** [review]
System test passed.

File: facade/nodeFacade.go [hunk 10]
```diff
@@ -213,7 +232,17 @@ func (nf *nodeFacade) GetTokenSupply(token string) (*apiData.ESDTSupply, error)
...
-	return nf.node.GetAllIssuedESDTs(tokenType)
+	ctx, cancel := nf.getContextForApiTrieRangeOperations()
+	defer cancel()
+
+	return nf.node.GetAllIssuedESDTs(tokenType, ctx)
+}
+
+func (nf *nodeFacade) getContextForApiTrieRangeOperations() (context.Context, context.CancelFunc) {
+	timeout := time.Duration(nf.wsAntifloodConfig.TrieOperationsDeadlineMilliseconds) * time.Millisecond
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply facade/nodeFacade.go]: done

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed

**Review by miiu96 — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

**Pull request merged**